dbms/src/Flash/Mpp/MPPTunnelSet.cpp

```diff
@@ -63,83 +63,189 @@ void MPPTunnelSetBase<Tunnel>::write(tipb::SelectResponse & response)
+                else
+                    tunnels[i]->write(compressed_tracked_packet->copy()); // NOLINT
+            }
         }
-        break;
     }
-    case CompressionMethod::LZ4:
+
+    if (is_broadcast)
     {
-        GET_METRIC(tiflash_exchange_data_bytes, type_hash_lz4_compression).Increment(sz);
-        break;
+        UPDATE_EXCHANGE_MATRIC(broadcast, CompressionMethod::NONE, local_tunnel_cnt * tracked_packet_bytes, local_tunnel_cnt * tracked_packet_bytes, true);
+        UPDATE_EXCHANGE_MATRIC(broadcast, compression_method, remote_tunnel_cnt * tracked_packet_bytes, remote_tunnel_cnt * compressed_tracked_packet_bytes, false);
     }
-    case CompressionMethod::ZSTD:
+    else
     {
-        GET_METRIC(tiflash_exchange_data_bytes, type_hash_zstd_compression).Increment(sz);
-        break;
-    }
-    default:
-        break;
+        UPDATE_EXCHANGE_MATRIC(passthrough, CompressionMethod::NONE, local_tunnel_cnt * tracked_packet_bytes, local_tunnel_cnt * tracked_packet_bytes, true);
+        UPDATE_EXCHANGE_MATRIC(passthrough, compression_method, remote_tunnel_cnt * tracked_packet_bytes, remote_tunnel_cnt * compressed_tracked_packet_bytes, false);
     }
 }
 
 template <typename Tunnel>
-void MPPTunnelSetBase<Tunnel>::broadcastOrPassThroughWrite(Blocks & blocks)
+void MPPTunnelSetBase<Tunnel>::broadcastOrPassThroughWrite(Blocks & blocks, bool is_broadcast)
 {
     RUNTIME_CHECK(!tunnels.empty());
     auto && tracked_packet = MPPTunnelSetHelper::ToPacketV0(blocks, result_field_types);
     if (!tracked_packet)
         return;
+    const size_t local_tunnel_cnt = std::accumulate(tunnels.begin(), tunnels.end(), 0, [](auto res, auto && tunnel) {
+        return res + tunnel->isLocal();
+    });
+
+    BroadcastOrPassThroughWriteImpl(
+        tunnels,
+        local_tunnel_cnt,
+        is_broadcast,
+        std::move(tracked_packet),
+        nullptr,
+        CompressionMethod::NONE);
+}
 
-    auto packet_bytes = tracked_packet->getPacket().ByteSizeLong();
-    checkPacketSize(packet_bytes);
-    // TODO avoid copy packet for broadcast.
-    for (size_t i = 1; i < tunnels.size(); ++i)
-        tunnels[i]->write(tracked_packet->copy());
-    tunnels[0]->write(std::move(tracked_packet));
+template <typename Tunnel>
+void MPPTunnelSetBase<Tunnel>::broadcastOrPassThroughWrite(Blocks & blocks, MPPDataPacketVersion version, CompressionMethod compression_method, bool is_broadcast)
+{
+    if (MPPDataPacketV0 == version)
+        return broadcastOrPassThroughWrite(blocks, is_broadcast);
+
+    RUNTIME_CHECK(!tunnels.empty());
+    const size_t local_tunnel_cnt = std::accumulate(tunnels.begin(), tunnels.end(), 0, [](auto res, auto && tunnel) {
+        return res + tunnel->isLocal();
+    });
+
+    size_t original_size = 0;
+    // encode by method NONE
+    auto && local_tunnel_tracked_packet = MPPTunnelSetHelper::ToPacket(std::move(blocks), version, CompressionMethod::NONE, original_size);
```

Comment:
Maybe we can also skip encode local_tunnel_tracked_packet if there is no local tunnel?

Reply:
The implementation of [CompressedWriteBuffer](https://github.com/pingcap/tiflash/blob/master/dbms/src/IO/CompressedWriteBuffer.cpp#L143) will write the original serialized data into its own `BufferWithOwnMemory` and [encode by CompressionMethod](https://github.com/pingcap/tiflash/blob/master/dbms/src/IO/CompressedWriteBuffer.cpp#L63-L66), then [write to next WriteBuffer](https://github.com/pingcap/tiflash/blob/master/dbms/src/IO/CompressedWriteBuffer.cpp#L135).

My implementation just separate above processes manually.
https://github.com/pingcap/tiflash/blob/4e84fa4db70f7de0f7fcc428c5a73c5bd4dfc865/dbms/src/Flash/Mpp/MPPTunnelSet.cpp#L240

Reply:
Yes, but the memory usage will be double or even triple if there is only one remote tunnel in this case, I prefer to not encode uncompressed packet if not needed.

Reply:
I mean if there is only one remote tunnel and the compression is not NONE